pkg/generator/secret_template_reconciler.go

```diff
@@ -0,0 +1,272 @@
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if secretTemplate.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	status := &reconciler.Status{
+		S:          secretTemplate.Status.GenericStatus,
+		UpdateFunc: func(st sgv1alpha1.GenericStatus) { secretTemplate.Status.GenericStatus = st },
+	}
+
+	status.SetReconciling(secretTemplate.ObjectMeta)
+	defer r.updateStatus(ctx, &secretTemplate)
+
+	res, err := r.reconcile(ctx, &secretTemplate)
+	//TODO is this overly defensive?
+	if err != nil {
+		secretTemplate.Status.Secret.Name = ""
+		if deleteErr := r.deleteChildSecret(ctx, &secretTemplate); deleteErr != nil {
+			return status.WithReconcileCompleted(res, deleteErr)
+		}
+	}
+
+	return status.WithReconcileCompleted(res, err)
+}
+
+func (r *SecretTemplateReconciler) reconcile(ctx context.Context, secretTemplate *sg2v1alpha1.SecretTemplate) (reconcile.Result, error) {
+	//Resolve input resources
+	inputResources, err := r.resolveInputResources(ctx, secretTemplate)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	//Template Secret Data
+	secretData := map[string][]byte{}
+	for key, expression := range secretTemplate.Spec.JSONPathTemplate.Data {
+		valueBuffer, err := JSONPath(expression).EvaluateWith(inputResources)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("templating data: %w", err)
+		}
+
+		decoded, err := base64.StdEncoding.DecodeString(valueBuffer.String())
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed decoding base64 from a Secret: %w", err)
+		}
+
+		secretData[key] = decoded
+	}
+
+	//Template Secret StringData
+	secretStringData := map[string]string{}
+	for key, expression := range secretTemplate.Spec.JSONPathTemplate.StringData {
+		valueBuffer, err := JSONPath(expression).EvaluateWith(inputResources)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("templating stringData: %w", err)
+		}
+
+		secretStringData[key] = valueBuffer.String()
+	}
+
+	//Create/Update Secret
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretTemplate.GetName(),
+			Namespace: secretTemplate.GetNamespace(),
+		},
+	}
+
+	if _, err = controllerutil.CreateOrUpdate(ctx, r.client, &secret, func() error {
+		secret.ObjectMeta.Labels = secretTemplate.Spec.JSONPathTemplate.Metadata.Labels
+		secret.ObjectMeta.Annotations = secretTemplate.Spec.JSONPathTemplate.Metadata.Annotations
+		secret.StringData = secretStringData
+		secret.Data = secretData
+
+		// Secret Type is immutable, so cannot be updated. TODO what to do here?
+		if secret.Type == "" {
+			secret.Type = secretTemplate.Spec.JSONPathTemplate.Type
+		}
+
+		return controllerutil.SetControllerReference(secretTemplate, &secret, scheme.Scheme)
+	}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("creating/updating secret: %w", err)
+	}
+
+	secretTemplate.Status.Secret.Name = secret.Name
+
+	return reconcile.Result{
+		RequeueAfter: syncPeriod,
```

Comment:
hmm, do we have any other ideas for the future how to make this more performant?

at least for when inputs are all secrets, can we only trigger reconciliation only based on watches?
for case where there are non-secret inputs, im not even sure how to do this best...

Reply:
Yeah this is an interesting problem and yes ideally we would be notified of changes to input resources through watches.

For arbitrary input resources this is made hard by the fact we use a separate Service Account for each SecretTemplate. This means that we cannot simply setup a single informer & watch for a GVK (since we have multiple identities and the watch call must use one identity) and on changes map to a corresponding SecretTemplate. So the options I see are:
1. Rely on period resync (as implemented in this PR)
1. Create a new watch and cached informer per GVK per SecretTemplate Service Account.

For option 2, given that would mean an explosion of watches to the API Server I think that may be even less scalable than the period resync. 

Have you explored this problem as part of https://github.com/vmware-tanzu/carvel-kapp-controller/ ? Seems very similar, user defined ServiceAccounts that access arbitrary Resources.

For Secrets it's a different matter because we use the same controller service account. This means we just need to watch for secrets and having a mapping function to the SecretTemplate.

Reply:
I spiked out just watching secrets here, https://github.com/Samze/carvel-secretgen-controller/pull/1/files we can do a follow up PR with these sort of changes.